Modules/Session/classes/class.ilObjSession.php

```diff
@@ -765,8 +813,144 @@ public function handleAutoFill()
 					}
 				}
 			}
-		}		
+		}
+	}
+
+	// cat-tms-patch start
+	/**
+	 * How should the tutors be configured?
+	 *
+	 * @param int $tutor_source
+	 */
+	public function setTutorSource($tutor_source) {
+		if($tutor_source !== self::TUTOR_CFG_MANUALLY
+			&& $tutor_source !== self::TUTOR_CFG_FROMCOURSE) {
+			throw new \InvalidArgumentException('ilObjSession::setTutorSource - invalid source: ' .$tutor_source);
+		}
+		$this->tutor_source = $tutor_source;
+	}
+
+	/**
+	 * How are the tutors configured?
+	 *
+	 * @return int
+	 */
+	public function getTutorSource() {
+		return $this->tutor_source;
+	}
+
+	/**
+	 * Get a list of users that are assigned as tutors at the course
+	 * this session lives in.
+	 *
+	 * @global type $tree
+	 * @return ilObjUser[]
+	 */
+	public function getParentCourseTutors() {
+		global $tree;
+		if(!$this->getRefId()) {
+			return array();
+		}
+		$parent = array(
+			'type'=>'none',
+			'ref_id' => $this->getRefId()
+		);
+		while($parent['type'] !== 'crs') {
+			$parent = $tree->getParentNodeData($parent['ref_id']);
+		}
```

Comment:
This might loop forever or crash, depending on what `$tree` thinks it should do when root is reached...